core/src/main/scala/core/builder/ServiceSpecValidator.scala

```diff
@@ -442,12 +437,23 @@ case class ServiceSpecValidator(
   private[this] def validateUnionTypeDiscriminatorValuesAreDistinct(): Seq[String] = {
     service.unions.flatMap { union =>
       dupsError(
-        s"Union[${union.name}] discriminator values",
-        union.types.map(getDiscriminatorValue)
+        s"Union[${union.name}] discriminator value",
+        getAllDiscriminatorValues(union),
       )
     }
   }
 
+  private[this] def getAllDiscriminatorValues(union: Union, resolved: Set[String] = Set.empty): Seq[String] = {
+    union.types.flatMap { t =>
+      service.unions.find(_.name == t.`type`) match {
+        case Some(nestedUnion) if !resolved.contains(nestedUnion.name) && union.name != nestedUnion.name => {
+          getAllDiscriminatorValues(nestedUnion, resolved ++ Seq(union.name))
```

Comment:
I think we need to add the discriminator values of all, including union types.

Reply:
not 100% sure I understand your comment here as this function should return ALL discriminator values (and the tests verify this). The specific check on union.name is to avoid an infinite loop, eg:

```
"user": {
  "types": [
    { "type": "user" }
  ] 
}
```

recursing here would result in an infinite loop. In this case, another validator will report an error of a circular definition